File: CS/DevExpressWalkthrough/ViewModels/Region/RegionViewModel.cs
using System;
using System.Linq;
using System.Linq.Expressions;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using DevExpress.Mvvm;
using DevExpress.Mvvm.POCO;
using DevExpressWalkthrough.Common.Utils;
using DevExpressWalkthrough.NorthwindEntitiesDataModel;
using DevExpressWalkthrough.Common.DataModel;
using DevExpressWalkthrough;
using DevExpressWalkthrough.Common.ViewModel;

namespace DevExpressWalkthrough.ViewModels {
    /// <summary>
    /// Represents the single Region object view model.
    /// </summary>
    public partial class RegionViewModel : SingleObjectViewModel<Region, int, INorthwindEntitiesUnitOfWork> {

        /// <summary>
        /// Creates a new instance of RegionViewModel as a POCO view model.
        /// </summary>
        /// <param name="unitOfWorkFactory">A factory used to create a unit of work instance.</param>
        public static RegionViewModel Create(IUnitOfWorkFactory<INorthwindEntitiesUnitOfWork> unitOfWorkFactory = null) {
            return ViewModelSource.Create(() => new RegionViewModel(unitOfWorkFactory));
        }

        /// <summary>
        /// Initializes a new instance of the RegionViewModel class.
        /// This constructor is declared protected to avoid undesired instantiation of the RegionViewModel type without the POCO proxy factory.
        /// </summary>
        /// <param name="unitOfWorkFactory">A factory used to create a unit of work instance.</param>
        protected RegionViewModel(IUnitOfWorkFactory<INorthwindEntitiesUnitOfWork> unitOfWorkFactory = null)
            : base(unitOfWorkFactory ?? UnitOfWorkSource.GetUnitOfWorkFactory(), x => x.Regions, x => x.RegionDescription) {
        }

        /// <summary>
        /// The view model for the RegionTerritories detail collection.
        /// </summary>
        public CollectionViewModel<Territory, string, INorthwindEntitiesUnitOfWork> RegionTerritoriesDetails
        {
            get { return GetDetailsCollectionViewModel((RegionViewModel x) => x.RegionTerritoriesDetails, x => x.Territories, x => x.RegionID, (x, key) => { x.RegionID = key; }); }
        }
    }
}

File: VB/DevExpressWalkthrough.VB/ViewModels/Region/RegionViewModel.vb
Imports System
Imports System.Linq
Imports System.Linq.Expressions
Imports System.Collections.Generic
Imports System.Collections.ObjectModel
Imports DevExpress.Mvvm
Imports DevExpress.Mvvm.POCO
Imports DevExpressWalkthrough.VB.Common.Utils
Imports DevExpressWalkthrough.VB.NorthwindEntitiesDataModel
Imports DevExpressWalkthrough.VB.Common.DataModel
Imports DevExpressWalkthrough.VB
Imports DevExpressWalkthrough.VB.Common.ViewModel
Namespace Global.DevExpressWalkthrough.VB.ViewModels
    ''' <summary>
    ''' Represents the single Region object view model.
    ''' </summary>
    Public Partial Class RegionViewModel
        Inherits SingleObjectViewModel(Of Region, Integer, INorthwindEntitiesUnitOfWork)
        ''' <summary>
        ''' Creates a new instance of RegionViewModel as a POCO view model.
        ''' </summary>
        ''' <param name="unitOfWorkFactory">A factory used to create a unit of work instance.</param>
        Public Shared Function Create(Optional ByVal unitOfWorkFactory As IUnitOfWorkFactory(Of INorthwindEntitiesUnitOfWork) = Nothing) As RegionViewModel
            Return ViewModelSource.Create(Function() New RegionViewModel(unitOfWorkFactory))
        End Function
        ''' <summary>
        ''' Initializes a new instance of the RegionViewModel class.
        ''' This constructor is declared protected to avoid undesired instantiation of the RegionViewModel type without the POCO proxy factory.
        ''' </summary>
        ''' <param name="unitOfWorkFactory">A factory used to create a unit of work instance.</param>
        Protected Sub New(Optional ByVal unitOfWorkFactory As IUnitOfWorkFactory(Of INorthwindEntitiesUnitOfWork) = Nothing)
            MyBase.New(If(unitOfWorkFactory, UnitOfWorkSource.GetUnitOfWorkFactory()), Function(ByVal x) x.Regions, Function(ByVal x) x.RegionDescription)
        End Sub
        ''' <summary>
        ''' The view model for the RegionTerritories detail collection.
        ''' </summary>
        Public ReadOnly Property RegionTerritoriesDetails As CollectionViewModel(Of Territory, String, INorthwindEntitiesUnitOfWork)
            Get
                Return GetDetailsCollectionViewModel(Function(ByVal x As RegionViewModel) x.RegionTerritoriesDetails, Function(ByVal x) x.Territories, Function(ByVal x) x.RegionID, Sub(ByVal x, ByVal key) x.RegionID = key)
            End Get
        End Property
    End Class
End Namespace
